girder/models/file.py

```diff
@@ -94,10 +95,19 @@ def download(self, file, offset=0, headers=True, endByte=None,
         :type extraParameters: str or None
         """
         if file.get('assetstoreId'):
-            return self.getAssetstoreAdapter(file).downloadFile(
+            fileDownload = self.getAssetstoreAdapter(file).downloadFile(
                 file, offset=offset, headers=headers, endByte=endByte,
                 contentDisposition=contentDisposition,
                 extraParameters=extraParameters)
+            if inspect.isgeneratorfunction(fileDownload):
+                def downloadGenerator():
+                    for data in fileDownload():
+                        yield data
+                    if endByte is None or endByte >= file['size']:
+                        events.trigger('download.complete', info={'id': file['_id']})
+                return downloadGenerator
+            else:
+                return fileDownload
```

Comment:
We should also add a try/except for cherrypy.redirect exceptions. We won't know for sure if the user completed the download, but we can log it, perhaps with a field that says `redirect: True` in the event info.